cloudinit/net/dhcp.py

```diff
@@ -364,4 +369,86 @@ def _trunc_error(cidr, required, remain):
     return static_routes
 
 
-# vi: ts=4 expandtab
+def kill_dhcp_client():
+    subp.subp(["pkill", "dhclient"], rcs=[0, 1])
+
+
+def clear_leases():
+    kill_dhcp_client()
+    files = glob.glob("/var/lib/dhcp/*")
+    for file in files:
+        os.remove(files)
+
+
+def start_service(dhcp_interface):
+    subp.subp(
+        ["service", "dhclient", "start", dhcp_interface],
+        rcs=[0, 1],
+        capture=True,
+    )
+
+
+def stop_service(dhcp_interface):
+    subp.subp(
+        ["service", "dhclient", "stop", dhcp_interface],
+        rcs=[0, 1],
+        capture=True,
+    )
+
+def get_dhclient_d():
+    # find lease files directory
+    supported_dirs = [
+        "/var/lib/dhclient",
+        "/var/lib/dhcp",
+        "/var/lib/NetworkManager",
+    ]
+    for d in supported_dirs:
+        if os.path.exists(d) and len(os.listdir(d)) > 0:
+            LOG.debug("Using %s lease directory", d)
+            return d
+    return None
+
+
+def get_latest_lease(lease_d=None):
+    # find latest lease file
+    if lease_d is None:
+        lease_d = get_dhclient_d()
+    if not lease_d:
+        return None
+    lease_files = os.listdir(lease_d)
+    latest_mtime = -1
+    latest_file = None
+
+    # lease files are named inconsistently across distros.
+    # We assume that 'dhclient6' indicates ipv6 and ignore it.
+    # ubuntu:
+    #   dhclient.<iface>.leases, dhclient.leases, dhclient6.leases
+    # centos6:
+    #   dhclient-<iface>.leases, dhclient6.leases
+    # centos7: ('--' is not a typo)
+    #   dhclient--<iface>.lease, dhclient6.leases
+    for fname in lease_files:
+        if fname.startswith("dhclient6"):
+            # avoid files that start with dhclient6 assuming dhcpv6.
+            continue
+        if not (fname.endswith(".lease") or fname.endswith(".leases")):
+            continue
+
+        abs_path = os.path.join(lease_d, fname)
+        mtime = os.path.getmtime(abs_path)
+        if mtime > latest_mtime:
+            latest_mtime = mtime
+            latest_file = abs_path
+    return latest_file
+
+
+def parse_dhcp_server_from_lease_file(lease_file):
```

Comment:
Is there opportunity to leverage `parse_dhcp_lease_file` here? 

We could also speed up this code potentially by reversing the order of leases processed to break on the first case of `dhcp-server-identifier` instead of processing the whole file and keeping the last case. It may not matter much in practice as the lease file processed in most cases may only have 1 lease in it though.

Reply:
> Is there opportunity to leverage parse_dhcp_lease_file here?

It looks like it. This code came from a different file and looks like a case of partial code duplication.

> We could also speed up this code potentially...

Agreed. I'd like for that to be a separate PR for a cleaner git history, since it doesn't fit the scope of this change.